everviz/plugins/config_editor/config_editor.py

```diff
@@ -0,0 +1,172 @@
+    """
+
+    def __init__(self, app, data_path, title="Configuration"):
+        super().__init__()
+        self.title = title
+        self.data_path = data_path
+        self.text_area = f"txt_area-{uuid4()}"
+        self.md_area = f"md_area-{uuid4()}"
+        self.editor_buttons = f"editor_buttons-{uuid4()}"
+        self.btn_edit = f"btn_edit-{uuid4()}"
+        self.btn_cancel = f"btn_cancel-{uuid4()}"
+        self.btn_upload = f"btn_upload-{uuid4()}"
+        self.btn_reset = f"btn_reset-{uuid4()}"
+        self.config_comp = f"config_comp-{uuid4()}"
+        self.editing = False
+        self.default_conf_path = Path(self.data_path).parent / "default"
+        self.default_conf = yaml.safe_load(self.get_config_str())
+        self.set_callbacks(app)
+
+    def get_config_str(self):
+        with open(self.data_path, "r") as f:
+            return "".join(f.readlines())
+
+    @staticmethod
+    def write_config(config, path):
+        with open(path, "w") as f:
+            yaml.safe_dump(config, f, default_flow_style=False, sort_keys=False)
+
+    def update_everviz_config(self, new_config_text):
+        new_config = yaml.safe_load(new_config_text)
+        old_config = yaml.safe_load(self.get_config_str())
+        if isinstance(new_config, dict) and old_config != new_config:
+            pages = new_config.get("pages", [])
+            if len(pages) > 0:
+                # Ensure the config editor page is always part of the config
+                conf_ed_content = {
+                    "ConfigEditor": {"data_path": self.data_path,},
+                }
+                # Check if the ConfigEditor plugin is part of the configuration
+                found = next(
+                    (page for page in pages if page["content"] == [conf_ed_content]),
+                    None,
+                )
+                if not found:
+                    pages.append(
+                        {"title": "Config editor", "content": [conf_ed_content]}
+                    )
+                if not self.default_conf_path.exists():
+                    self.write_config(self.default_conf, self.default_conf_path)
+                self.write_config(new_config, self.data_path)
+
+    def render_text_component(self):
+        config_str = self.get_config_str()
+        if self.editing:
+            style = {"width": "100%", "height": 800}
+            return [dcc.Textarea(id=self.text_area, value=config_str, style=style,)]
+
+        md_content = f"""```
+{config_str}
+```
+"""
+        return [
+            dcc.Textarea(id=self.text_area, value=config_str, hidden=not self.editing,),
+            dcc.Markdown(id=self.md_area, children=md_content,),
+        ]
+
+    def render_buttons(self):
+        if not self.editing:
+            return [
+                html.Button("Edit", id=self.btn_edit),
+                # Had to wrap the button in a div because the hidden option was not
+                # working on the button element
```

Comment:
Do you know if it is possible to deactivate buttons instead of hiding them? That might be an alternative.

Reply:
yes, it is possible to deactivate buttons but that did not work as expected either. I think I tried 100 things before I ended with the current version of the code dash fought me hard along the way.

Reply:
yes, I noted that dash is not always a pleasure...